### PR TITLE
Verify the OPAL compatibility with XFS features (#1997832)

### PR DESCRIPTION
**Pull request by poncovka**
Shows a warning on ppc64le systems with old petitboot, when the /boot
filesystem is formatted with XFS. The system won't be bootable.

Resolves: rhbz#1997832

### CODE REVIEW

**Review by VladimirSlavik — APPROVED** [review]
Looks good to me. Thank you!

**Review by M4rtinK — APPROVED** [review]
Looks good to me as well, thanks a lot! :)

**Review by jkonecny12 — APPROVED** [review]
Looks great to me. Thanks!
Could you please attach an image of the warning here?

File: pyanaconda/modules/storage/checker/utils.py [hunk 1]
```diff
@@ -225,6 +227,52 @@ def verify_gpt_biosboot(storage, constraints, report_error, report_warning):
...
+    """ Verify the petitboot version compatibility.
+
+    :param storage: a storage to check
+    :param constraints: a dictionary of constraints
+    :param report_error: a function for error reporting
+    :param report_warning: a function for warning reporting
+    """
+    if arch.get_arch() == "ppc64le" and os.access("/sys/firmware/opal", os.F_OK):
+        # Check the version of the petitboot loader.
+        version = _get_petitboot_version()
```
Comment (jstodola) [inline pyanaconda/modules/storage/checker/utils.py:240]: I don't think we care about the petitboot version at all. We need to ensure that the firmware kernel version (for OPAL systems) is 5.10 or newer. The kernel version can be found in `/proc/device-tree/ibm,firmware-versions/linux` or by executing `lsmcode` - see bug [1977236](https://bugzilla.redhat.com/show_bug.cgi?id=1977236).
@sharkcz , can you please confirm?

**Review by jstodola — CHANGES_REQUESTED** [review]

File: pyanaconda/modules/storage/checker/utils.py (hunk 1 above)
Comment (sharkcz) [reply pyanaconda/modules/storage/checker/utils.py:240]: - @jstodola is right, it's kernel version in the OPAL firmware that matters
- I would call the function/method `verify_opal_compatibility`
- the check for arch and `/sys/firmware/opal` could be switched to `if get_platform() == 'PowerNV'`
- if you can't positively detect the kernel version, then I would assume it's "bad/old"

Comment (poncovka) [reply pyanaconda/modules/storage/checker/utils.py]: Thanks for the pointers! I have got confused by https://bugzilla.redhat.com/show_bug.cgi?id=1997832#c2. Hopefully, it is fine now.

File: pyanaconda/modules/storage/checker/utils.py [hunk 2]
```diff
@@ -225,6 +226,50 @@ def verify_gpt_biosboot(storage, constraints, report_error, report_warning):
...
+                "Your firmware doesn't support XFS file system features "
+                "on the /boot file system. The system will not be bootable. "
+                "Please, upgrade the firmware or change the file system type."
+            ))
+
+
+def _get_opal_firmware_kernel_version():
+    """Get the firmware kernel version for OPAL systems.
+
+    For example: 5.10.50-openpower1-p59fd803
```
Comment (jstodola) [inline pyanaconda/modules/storage/checker/utils.py:256]: Please, be aware that the version could also be something like "v4.15.9-openpower1-p9e03417" (see https://bugzilla.redhat.com/show_bug.cgi?id=1977236#c16) - in that case comparing of versions would throw a traceback:
```
>>> LooseVersion("5.10") > LooseVersion("v4.15.9-openpower1-p9e03417")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.9/distutils/version.py", line 64, in __gt__
    c = self._cmp(other)
  File "/usr/lib64/python3.9/distutils/version.py", line 341, in _cmp
    if self.version < other.version:
TypeError: '<' not supported between instances of 'int' and 'str'
```

Comment (sharkcz) [reply pyanaconda/modules/storage/checker/utils.py:256]: I think there are 2 options, either check the string from `ibm,firmware-versions/linux` against `<num>.<num>.<num>-something` format or wrap the comparison in a `try` block. This is what I meant with the positive identification of 5.10+ kernel. If the installer can't be sure about the version, then it's safer to expect system without new-enough XFS support.

Comment (poncovka) [reply pyanaconda/modules/storage/checker/utils.py:272]: I have put back the code that removes `v` from the version and moved the version comparison into a try-except block.

File: tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py [hunk 3]
```diff
@@ -108,3 +115,117 @@ def lvm_verification_test(self):
...
+        mocked_arch.get_arch.return_value = "ppc64le"
+        mocked_arch.is_powernv.return_value = False
+
+        verify_opal_compatibility(None, {}, None, reporter)
+        reporter.assert_not_called()
+
+    @patch("pyanaconda.modules.storage.checker.utils._get_opal_firmware_kernel_version")
+    @patch("pyanaconda.modules.storage.checker.utils.arch")
+    def opal_verification_version_test(self, mocked_arch, version_getter):
+        """Check verify_opal_compatibility with a higher firmware version."""
```
Comment (VladimirSlavik) [inline tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py:176]: I don't think the test name matches the docstring.

File: tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py [hunk 4]
```diff
@@ -108,3 +115,117 @@ def lvm_verification_test(self):
...
+        version_getter.return_value = "5.10.50-openpower1-p59fd803"
+
+        verify_opal_compatibility(None, {}, None, reporter)
+        reporter.assert_not_called()
+
+    @patch.object(XFS, "mountable", new_callable=PropertyMock)
+    @patch("pyanaconda.modules.storage.checker.utils._get_opal_firmware_kernel_version")
+    @patch("pyanaconda.modules.storage.checker.utils.arch")
+    def opal_verification_test(self, mocked_arch, version_getter, xfs_mountable):
+        """Check verify_opal_compatibility."""
```
Comment (VladimirSlavik) [inline tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py:189]: And this test is actually for a too low version, but we can't learn that here.

**Review by VladimirSlavik — COMMENTED** [review]
I wonder what version strings are out there in the wild, expecting just "v" makes me nervous.

The code outgrew the test names and docstrings, I guess, but that's a minor thing.

File: tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py (hunk 3 above)
Comment (poncovka) [reply tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py]: Fixed.

File: tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py (hunk 4 above)
Comment (poncovka) [reply tests/nosetests/pyanaconda_tests/modules/storage/module_storage_checker_test.py]: Fixed.

**Comment by poncovka** [comment]
> I wonder what version strings are out there in the wild, expecting just "v" makes me nervous.

The current code should be able to survive any other weird strings that and show the warning as suggested in https://github.com/rhinstaller/anaconda/pull/3598#discussion_r708975700.

**Review by VladimirSlavik — APPROVED** [review]
Thank you!

I agree that the error message is as good as it can get without actual data about past versions.

**Comment by poncovka** [comment]
Screenshot from 2021-09-15 16-40-58

**Comment by sharkcz** [comment]
Looks good, thanks. Ideally we could a link to a KBase article to the message, but it hasn't been written yet (but is planned).
How does the automatic partitioning work? Does it default to ext4 or does it require attention from the user?

**Comment by poncovka** [comment]
> How does the automatic partitioning work? Does it default to ext4 or does it require attention from the user?

With the automatic partitioning, the status of the Installation Destination spoke in the Summary hub informs about warnings in the storage configuration. It is necessary to reenter the Installation Destination spoke to see the warning.

Screenshot from 2021-09-13 14-24-53

**Review by jstodola — APPROVED** [review]
I have not tried any manual testing, but I cannot see any other possible issues. Thanks!

**Comment by sharkcz** [comment]
OK, and for kickstart installation it means manually setting up the disk layout and not relying on on `autopart`, right? I am thinking how this can affect beaker provisioning and similar situation ...

**Comment by poncovka** [comment]
> OK, and for kickstart installation it means manually setting up the disk layout and not relying on on `autopart`, right? I am thinking how this can affect beaker provisioning and similar situation ...

It is possible to use `autopart --fstype ext4`. Just a note, the warning doesn't block the installation. It just informs the user about the issue, but it is possible to ignore the warning and continue with the current settings. If we want to enforce the change, we need to report an error from the verification function. However, @jkonecny12 mentioned that the warning is sufficient in this case.

**Comment by jkonecny12** [comment]
Wrote a comment on the bug to make it totally clear that we are just adding non-blocking warning. To avoid confusion.

**Comment by sharkcz** [comment]
This should go to the master branch as well, the problem is not specific to RHEL, it affects Fedora too.

**Comment by VladimirSlavik** [comment]
> This should go to the master branch as well, the problem is not specific to RHEL, it affects Fedora too.

No worries! The `port to Fedora` label means exactly that.

**Comment by sharkcz** [comment]
Thanks :-)

**Comment by poncovka** [comment]
Ported to upstream in https://github.com/rhinstaller/anaconda/pull/3620.

**Pull request merged**